pdns/misc.cc

```diff
@@ -198,6 +198,30 @@ size_t writen2WithTimeout(int fd, const void * buffer, size_t len, const struct
   return len;
 }
 
+auto pdns::getMessageFromErrno(const int errnum) -> std::string
+{
+  const size_t errLen = 2048;
+  std::string errMsgData{};
+  errMsgData.reserve(errLen);
```

Comment:
Ah. `resize()` and `size()` are the correct methods to use here, I don't know why I went for `reserve()` which would leave other string metadata (like the size/length) incorrect.

Reply:
However, regarding avoiding the copy, I don't think that's possible. From `strerror_r(3)`:

> The GNU-specific strerror_r() returns a pointer to a string containing the error message.  This may be either a pointer to a string that the function stores in buf, or a  pointer  to  some
>        (immutable)  static  string (in which case buf is unused).  If the function stores a string in buf, then at most buflen bytes are stored (the string may be truncated if buflen is too small
>        and errnum is unknown).  The string always includes a terminating null byte ('\0').

Reply:
Right, I always forget that the GNU version tries to do the optimization for you.. I guess in the GNU case we could compare `errMsg` to `errMsgData.data()`, though, since we have already allocated a buffer, and `assign()` the content pointed to by `errMsg` if needed, but I'm not sure it is worth it.
It looks like the XSI case is not going to work if `res != 0`, though?

Reply:
>  It looks like the XSI case is not going to work if res != 0, though?

Ah, I think you mean if `res == 0`?

Reply:
Uh, yes, sorry!